### PR TITLE
resolved_ts: add resolved ts advancer and event observer

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
Separate from #9435
Introduce resolved ts advancer and event observer.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
* N/A

### CODE REVIEW

**Review by NingLin-P — COMMENTED** [review]
The rest LGTM

File: components/resolved_ts/src/observer.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    fn on_flush_apply(&self, engine: E) {
+        if !self.cmd_batches.borrow().is_empty() {
+            let batches = self.cmd_batches.replace(Vec::default());
+            let mut region = Region::default();
+            region.mut_peers().push(Peer::default());
+            // Create a snapshot here for preventing the old value was GC-ed.
+            let snapshot =
+                RegionSnapshot::from_snapshot(Arc::new(engine.snapshot()), Arc::new(region));
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply components/resolved_ts/src/observer.rs:54]: Yes, it is only needed after enabling old value. I'm considering a proper way to control this, added a TODO here, will address this later in a separate PR.

**Review by NingLin-P — APPROVED** [review]
LGTM

File: components/resolved_ts/src/observer.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        self.cmd_batches
+            .borrow_mut()
+            .push(CmdBatch::new(observe_id, region_id));
+    }
+
+    fn on_apply_cmd(&self, observe_id: ObserveID, region_id: u64, cmd: Cmd) {
+        self.cmd_batches
+            .borrow_mut()
+            .last_mut()
+            .expect("should exist some cmd batch")
```
Comment (overvenus) [inline components/resolved_ts/src/observer.rs:43]: Could you attach region id in the panic log?

File: components/resolved_ts/src/observer.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    fn on_flush_apply(&self, engine: E) {
+        if !self.cmd_batches.borrow().is_empty() {
+            let batches = self.cmd_batches.replace(Vec::default());
+            let mut region = Region::default();
+            region.mut_peers().push(Peer::default());
+            // Create a snapshot here for preventing the old value was GC-ed.
+            // TODO: only need it after enabling old value, may add a flag to indicate whether to get it.
+            let snapshot =
+                RegionSnapshot::from_snapshot(Arc::new(engine.snapshot()), Arc::new(region));
+            if let Err(e) = self.scheduler.schedule(Task::ChangeLog {
```
Comment (overvenus) [inline components/resolved_ts/src/observer.rs:56]: Is there any test covering it?

Comment (5kbpers) [reply components/resolved_ts/src/observer.rs:56]: Will cover it in later integration tests

File: components/resolved_ts/src/endpoint.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    ChangeLog {
+        cmd_batch: Vec<CmdBatch>,
+        snapshot: RegionSnapshot<S>,
+    },
+    // ScanLocks {
+    //     region_id: u64,
+    //     observe_id: ObserveID,
+    //     entry: ScanEntry,
+    // },
```
Comment (overvenus) [inline components/resolved_ts/src/endpoint.rs:40]: When do we add the variant?

**Review by overvenus — APPROVED** [review]
LGTM

File: components/resolved_ts/src/endpoint.rs (hunk 4 above)
Comment (5kbpers) [reply components/resolved_ts/src/endpoint.rs:40]: Will be added in the next PR.

**Pull request merged**